src/librustc_typeck/astconv.rs

```diff
@@ -648,7 +656,71 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                                 if arg_count.correct.is_ok()
                                     && arg_count.explicit_late_bound == ExplicitLateBound::No
                                 {
-                                    Self::generic_arg_mismatch_err(tcx.sess, arg, kind.descr());
+                                    // We're going to iterate over the parameters to sort them out, and
+                                    // show that order to the user as a possible order for the parameters
+                                    let mut ordered_params = defs.params.clone();
```

Comment:
Instead of two very similar match conditions, could we do something like the following?
```rust
let (ordered_params, mut params_present) = defs.params.clone().map(|param| (param, match param.kind { … })).sort_by_key(|(_, ord)| ord);
params_present.dedup();
```

Reply:
This was marked as resolved, but we're still matching on `param.kind` twice?

Reply:
Yeah the code mentioned there doesn't seem to compile. It seems to compile something about not being able to convert from map to tuple

Reply:
I couldn't quite think off the top of my head what the types would be, so it may require a little adaption, but I think it would be good to only match on `param.kind` once if that results in cleaner code.

Reply:
Alright, cool. I'll come up with something

Reply:
Hey @varkor. I'm getting this error:
```
the trait `std::hash::Hash` is not implemented for `rustc_middle::ty::GenericParamDef`
```
when trying to insert a GenericParamDef into a hashset.
Any idea how I can overcome this?

Reply:
My idea was not to use a hash set any more: we're using a vector instead, but because it's sorted, once we `dedup`, the elements will be unique.